### ISSUE TITLE
ESI requests break persistent remember me tokens

### ISSUE THREAD

**Issue by leofeyer** [symfony/symfony#31078]
We are using the [`PersistentTokenBasedRememberMeServices`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Http/RememberMe/PersistentTokenBasedRememberMeServices.php) service together with the [`DoctrineTokenProvider`](https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Doctrine/Security/RememberMe/DoctrineTokenProvider.php) class to manage "remember me" tokens. The idea behind the service is to update the token hash after each usage, so the same token cannot be used twice.

Now, if you have a "remember me" cookie and request a page that includes an ESI fragment, both the main request and the ESI request will trigger the auto-login process. However, after the main request has used the token, the hash is changed and the cookie is regenerated. By the time the ESI request is processed, the "remember me" token can no longer be verified (CookieTheftException) and the cookie is deleted in the [`loginFail`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Http/RememberMe/AbstractRememberMeServices.php#L163) method.

If the user was successfully authenticated based on the "remember me" cookie in the main request, the response will contain the updated "remember me" cookie as well as a session cookie. Should these cookies not be passed to the ESI request?

**Comment by leofeyer** [comment]
Related: symfony/symfony#27910 @paragonie-scott

**Comment by stof** [comment]
No, they should not, as that's not how ESI works when using Varnish or other reverse proxies. Reading response cookies to update the cookies sent in subsequent requests is something done by the browser, not by ESI-capable reverse proxies.

**Comment by stof** [comment]
#27910 is not related to this. Storing a hash of the value rather than the value itself is entirely orthogonal to the fact of changing the value on each usage.

**Comment by leofeyer** [comment]
Ok, but then the ESI request will trigger the auto-login process again and delete the "remember me" cookie (see above). So you are implying that we cannot use persistent tokens together with ESI, aren't you?

**Comment by leofeyer** [comment]
Any updates on this one? @nicolas-grekas Can you confirm that Symfony intentionally does not support persistent rememberme tokens with series in combination with ESI?

**Comment by nicolas-grekas** [comment]
> Should these cookies not be passed to the ESI request?

as @stof explained, Varnish doesn't do that. Maybe you could make it do so but that's out of scope.
If you do, that'd mean each ESI request would have a different rememberme cookie, since they can be used only once?
This wouldn't work because there would be race conditions. Your requirement of having cookies valid once mandates making requests on after the other. That'd defeat a significant benefit of ESI.

Base on your description above, I would look for a way to accept already consumed token from ESI responses (for some duration). Since they're signed, that'd be safe.

All that is on you, nothing we can do in Symfony to me.

**Comment by leofeyer** [comment]
Agreed. Thank you for your explanation. 😊

**Comment by aschempp** [comment]
Also see https://github.com/symfony/symfony/issues/18384
